### PR TITLE
[export] Add effect token to export

### PR DESCRIPTION
**Pull request by angelayi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #121552
* __->__ #121424

Following the creation of effect tokens (https://github.com/pytorch/pytorch/pull/120296), we want to now add support for these tokens in export because the calling/returning convention has changed. The inputs are now `(tokens, params, buffers, constants, user_inputs)` and the outputs are `(tokens, buffer_mutations, user_mutations, user_outputs)`. The graph looks something like: 
```
graph():
    %arg0_1 : [num_users=1] = placeholder[target=arg0_1]
    %attr : [num_users=2] = placeholder[target=attr]
    %arg1_1 : [num_users=2] = placeholder[target=arg1_1]
    %with_effects : [num_users=2] = call_function[target=torch._higher_order_ops.effects.with_effects](args = (%arg0_1, _TorchScriptTesting.takes_foo.default, %attr, %arg1_1), kwargs = {})
    %getitem : [num_users=1] = call_function[target=operator.getitem](args = (%with_effects, 0), kwargs = {})
    %getitem_1 : [num_users=1] = call_function[target=operator.getitem](args = (%with_effects, 1), kwargs = {})
    %with_effects_1 : [num_users=2] = call_function[target=torch._higher_order_ops.effects.with_effects](args = (%getitem, _TorchScriptTesting.takes_foo.default, %attr, %getitem_1), kwargs = {})
    %getitem_2 : [num_users=1] = call_function[target=operator.getitem](args = (%with_effects_1, 0), kwargs = {})
    %getitem_3 : [num_users=1] = call_function[target=operator.getitem](args = (%with_effects_1, 1), kwargs = {})
    %add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%arg1_1, %getitem_3), kwargs = {})
    return (getitem_2, add)
```

During unlifting, we will first remove the tokens and with_effect calls using the `remove_effect_tokens` pass. (cc @SherlockNoMad on the pass to remove tokens). This is so that this won't change the calling conventions when retracing. The graph after unlifting looks something like:
```
graph():
    %attr_1 : [num_users=2] = get_attr[target=attr]
    %arg1_1 : [num_users=2] = placeholder[target=arg1_1]
    %takes_foo_default_1 : [num_users=1] = call_function[target=torch.ops._TorchScriptTesting.takes_foo.default](args = (%attr_1, %arg1_1), kwargs = {})
    %takes_foo_default : [num_users=1] = call_function[target=torch.ops._TorchScriptTesting.takes_foo.default](args = (%attr_1, %takes_foo_default_1), kwargs = {})
    %add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%arg1_1, %takes_foo_default), kwargs = {})
    return (add,)
```

Serialization support will be added in a followup.
Note: tokens only affect custom ops that take in ScriptObjects, not ScriptObject methods yet.

Differential Revision: [D54639390](https://our.internmc.facebook.com/intern/diff/D54639390)

### CODE REVIEW

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54639390).

File: torch/export/_remove_effect_tokens_pass.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+from .graph_signature import (
+    InputKind,
+    InputSpec,
+    OutputKind,
+    OutputSpec,
+    TensorArgument,
+)
+
+
+def remove_effect_tokens(ep: ExportedProgram) -> ExportedProgram:
```
Comment (zhxchen17) [inline torch/export/_remove_effect_tokens_pass.py:16]: To make it more obviously private.
```suggestion
def _remove_effect_tokens(ep: ExportedProgram) -> ExportedProgram:
```

**Review by zhxchen17 — COMMENTED** [review]
export part looks fine to me, small nit

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54639390).

File: test/export/test_torchbind.py [hunk 2]
```diff
@@ -111,7 +111,59 @@ def __init__(self):
...
+
+        m = MyModule()
+        input = torch.ones(2, 3)
+        with enable_torchbind_tracing():
+            ep = torch.export.export(m, (input,), strict=False)
+
+        unlifted = ep.module()
+        self.assertEqual(m(input), unlifted(input))
+
+        with enable_torchbind_tracing():
```
Comment (tugsbayasgalan) [inline test/export/test_torchbind.py:126]: Are we always expected to call this context manager whenever there is TorchBind object in the graph?

**Review by tugsbayasgalan — APPROVED** [review]
Nice! Could we also add this to assert nodes in the graph? (as a follow up)

File: test/higher_order_ops/test_with_effects.py [hunk 3]
```diff
@@ -86,11 +85,13 @@ def forward(self, x):
         self.assertExpectedInline(
             str(gm.code).strip(),
             """\
-def forward(self, arg0_1):
+def forward(self, arg0_1, arg1_1):
     _tensor_constant0 = self._tensor_constant0
-    takes_foo = torch.ops._TorchScriptTesting.takes_foo.default(_tensor_constant0, arg0_1);  _tensor_constant0 = None
-    add = torch.ops.aten.add.Tensor(arg0_1, takes_foo);  arg0_1 = takes_foo = None
-    return (add,)""",  # noqa: B950
+    with_effects = torch._higher_order_ops.effects.with_effects(arg0_1, torch.ops._TorchScriptTesting.takes_foo.default, _tensor_constant0, arg1_1);  arg0_1 = _tensor_constant0 = None
```
Comment (zou3519) [inline test/higher_order_ops/test_with_effects.py:90]: I'm confused --  how did we get this working without Yidi's PR over at https://github.com/pytorch/pytorch/pull/120045? IMO if the there's no abstract impl for a class, then exporting should error out

File: torch/export/_unlift.py [hunk 4]
```diff
@@ -273,6 +274,7 @@ def _create_stateful_graph_module(
 
 
 def _unlift_exported_program_lifted_states(ep: ExportedProgram) -> torch.nn.Module:
+    ep = _remove_effect_tokens(ep)
```
Comment (zou3519) [inline torch/export/_unlift.py:277]: What is "unlifting" and where does it happen?

File: test/higher_order_ops/test_with_effects.py (hunk 3 above)
Comment (angelayi) [reply test/higher_order_ops/test_with_effects.py:90]: This is just with aot_export_module, aka suo's old hacks

File: torch/export/_unlift.py (hunk 4 above)
Comment (angelayi) [reply torch/export/_unlift.py:277]: Unlifting basically takes the exported program and returns a torch.fx.GraphModule with the parameters/buffers unlifted from inputs and as get_attr nodes, and any mutations put back into the graph with copy_. Users can call that eagerly and do what they would do to GraphModules returned by symbolic_trace.

Comment (zou3519) [reply torch/export/_unlift.py:277]: sg!

File: test/export/test_torchbind.py (hunk 2 above)
Comment (angelayi) [reply test/export/test_torchbind.py:126]: For now, with suo's hacks yes. But I think Yidi's working on a legit thing

**Pull request closed without merging**